### PR TITLE
Refactor scripting

### PR DESCRIPTION
**Pull request by felipemanga**
TL;DR: This lays the groundwork necessary for issue #169.
Sorry this PR isn't in small, easy-to-chew chunks, but it wouldn't make sense to split it up.

The goal is to be able to expand the scripting API so that it can be used for more things: 
- custom file format loaders/exporters
- new tools, filters, menu options and dialogs
- procedural art/animation
- support for other scripting languages

For this growth to be done in a sustainable way, some new concepts had to be introduced:
- Dependency Injection: This is a design pattern that makes it easier to add functionality while keeping the code modular and flexible.

- script::ScriptObject: These are classes located in `src/app/script/api`. They are responsible for exposing Libresprite functionality to scripting engines without being specific to any one engine. This neutrality is important as it allows multiple engines to be integrated without increasing the amount of work/code exponentially.

- script::Value: this class is similar to std::any or std::variant, but provides more flexibility in that we can add conversions as-needed (string "42" into int 42, for example). This is necessary because scripting engines have their own types and existing LibreSprite code will likely not match those types. script::Value provides a common set of types that everyone can agree on.

- script::Function: This is similar to an std::function, but instead of calling it with a set of parameters (`func(a, b, c)`), it uses an array: `func.arguments = {a, b, c}; func();`. With this it is possible for a scripting engine to load up as many arguments as necessary, and the function in the end doesn't need to have any special signature or handle arguments in an unusual way. 

## Future work
- Allow scripts to open dialog windows with input boxes and buttons for tweaking the script's parameters
- The selection_class has been disabled for now.
- Allow scripts to add file format encoders/decoders
- Support for new engines: v8? Python? Lua?

### CODE REVIEW

File: src/script/engine_delegate.h [hunk 1]
```diff
@@ -1,21 +1,20 @@
...
-  class EngineDelegate {
-  public:
-    virtual ~EngineDelegate() { }
-    virtual void onConsolePrint(const char* text) = 0;
-  };
+    class EngineDelegate : public Injectable<EngineDelegate> {
+    public:
+        virtual ~EngineDelegate() { }
+        virtual void onConsolePrint(const char* text) = 0;
+    };
```
Comment (Zughy) [inline src/script/engine_delegate.h:18]: indentation was correct

File: src/app/script/api/cel_script.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+                [this](int x){m_cel->setPosition(x, m_cel->y()); return x;});
+
+    addProperty("y",
+                [this]{return m_cel->y();},
+                [this](int y){m_cel->setPosition(m_cel->x(), y); return y;});
+
+    addProperty("image", [this]{return m_image.get();});
+
+    addProperty("frame", [this]{return m_cel->frame();});
+
```
Comment (Zughy) [inline src/app/script/api/cel_script.cpp:27]: Spaces between functions not needed (also because other files don't have it)

File: src/app/script/api/layer_script.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                  m_layer->setName(name);
+                  return name;
+                });
+
+    addProperty("isImage", [this]{return m_layer->isImage();});
+
+    addProperty("isBackground", [this]{return m_layer->isBackground();});
+
+    addProperty("isTransparent", [this]{return m_layer->isTransparent();});
+
```
Comment (Zughy) [inline src/app/script/api/layer_script.cpp:26]: same as before

File: src/app/script/api/sprite_script.cpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    addProperty("height",
+                [this]{return m_sprite->height();},
+                [this](int height){
+                  transaction().execute(new app::cmd::SetSpriteSize(m_sprite, m_sprite->width(), height));
+                  return 0;
+                });
+
+    addProperty("colorMode", [this]{ return m_sprite->pixelFormat();});
+
+    addProperty("selection", [this]{ return this; });
```
Comment (Zughy) [inline src/app/script/api/sprite_script.cpp:49]: same

**Comment by Zughy** [comment]
As said on Discord/Matrix, I'm nowhere near having OP's coding knowledge, so I can't really evaluate this PR in full. Nevertheless, he was so kind he guided me through the code when answering my questions, allowing me to understand at least the general structure. Thus I'll evaluate at least that: 
1. the idea of easily support more languages is great
2. confusing class names like `ImageWrap` are gone: good
3. As someone who's struggled understanding the code, I'd like to see a few comments here and there to give direction. I understand I'm below the average, so I'm not asking for a detailed explanation of what happens. I'm simply inclined to think this isn't the easiest thing to grasp in general, so just a pinch of info would be great (e.g. `injection.h`)
4. Tomorrow I'll compile and test a script to see if it works smoothly

**Comment by Zughy** [comment]
- Script works
- docs added

Fine for me

**Review by maujin111 — APPROVED** [review]
